packages/installer/src/transforms/prisma/print-schema.ts

```diff
@@ -0,0 +1,140 @@
+import {DataSource, DMMF, GeneratorConfig} from "@prisma/generator-helper"
+import {ConfigMetaFormat, IntrospectionEngine} from "@prisma/sdk"
+
+/**
+ * Takes the schema.prisma document parsed from @prisma/sdk and serializes it
+ * back to a schema.prisma source string. Since the SDK doesn't parse the whole
+ * data model, we also execute the IntrospectionEngine to fill in the remaining
+ * details and prettify the source code.
+ *
+ * @param doc - the document containing data model
+ * @param config - the configuration containing the data sources and generators
+ * @returns the schema.prisma source string
+ */
+export async function printSchema(doc: DMMF.Document, config: ConfigMetaFormat): Promise<string> {
+  let aspects = []
+  aspects.push(...config.datasources.map(printDatasource))
+  aspects.push(...config.generators.map(printGenerator))
+  aspects.push(...doc.datamodel.models.map(printModel))
+  aspects.push(...doc.datamodel.enums.map(printEnum))
+  const schema = aspects.filter(Boolean).join("\n")
+
+  // The IntrospectionEngine can examine the existing data model and add any
+  // aspects that are missing from the schema document, such as indexes, but
+  // there is no database to inspect when running tests so we'll skip it for
+  // now.
+  if (process.env.NODE_ENV === "test") return schema
+
+  const engine = new IntrospectionEngine()
+  const {datamodel, warnings} = await engine.introspect(schema, false)
+
+  // TODO: something besides console.warn() here
+  if (warnings && warnings.length > 0) console.warn(warnings)
+  return datamodel
+}
+
+function printDatasource({name, activeProvider, url}: DataSource) {
+  const children = [
+    activeProvider && `provider = "${activeProvider}"`,
+    url && url.fromEnvVar ? `url      = env("${url.fromEnvVar}")` : `url      = "${url.value}"`,
+  ]
+    .filter(Boolean)
+    .join("\n  ")
+
+  return `datasource ${name} {
+  ${children}
+}`
+}
+
+function printGenerator({name, provider, output, binaryTargets, previewFeatures}: GeneratorConfig) {
+  const children = [
+    provider && `provider        = "${provider}"`,
```

Comment:
These extra spaces before the key look a little odd to me.  Any reason for that?

Reply:
I think the idea is you are trying to line up the generated code on the `=` sign and `previewFeatures` is the largest key, but that might not be there.  I'm wondering if we want to keep things simple and leave the auto formatting up to a different process / keep it manual.  If not, we'd need to make this piece a little smarter.

Reply:
👍 Yeah, purpose of that code was to line up the equals signs, but the whole document gets prettified by the IntrospectionEngine so it isn't necessary. I don't think `printSchema()` needs to be tested independently from `produceSchema()` (the callback function can be a no-op) because it's much more convenient to write your test input as a string than it would be to build out DMMF objects manually.